modules/llrt_crypto/src/subtle/mod.rs

```diff
@@ -0,0 +1,404 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+mod decrypt;
+mod derive_bits;
+mod digest;
+mod encrypt;
+mod generate_key;
+mod sign;
+mod verify;
+
+use decrypt::decrypt;
+use derive_bits::derive_bits;
+use digest::digest;
+use encrypt::encrypt;
+use generate_key::generate_key;
+use sign::sign;
+use verify::verify;
+
+use aes::{cipher::typenum::U16, Aes256};
+use aes_gcm::AesGcm;
+use hmac::Hmac;
+use llrt_utils::{bytes::ObjectBytes, object::ObjectExt, result::ResultExt};
+use rquickjs::{Array, ArrayBuffer, Ctx, Exception, Result, Value};
+use sha2::Sha256;
+
+pub type HmacSha256 = Hmac<Sha256>;
+pub type Aes256Gcm = AesGcm<Aes256, U16>;
+
+#[derive(Debug)]
+pub enum Sha {
+    Sha1,
+    Sha256,
+    Sha384,
+    Sha512,
+}
+
+fn get_sha(ctx: &Ctx<'_>, hash: &str) -> Result<Sha> {
+    match hash.to_ascii_uppercase().as_str() {
+        "SHA-1" => Ok(Sha::Sha1),
+        "SHA-256" => Ok(Sha::Sha256),
+        "SHA-384" => Ok(Sha::Sha384),
+        "SHA-512" => Ok(Sha::Sha512),
+        _ => Err(Exception::throw_message(ctx, "hash not found")),
+    }
+}
+
+#[derive(Debug)]
+pub enum Algorithm {
+    Hmac,
+    AesGcm(Vec<u8>),
+    AesCbc(Vec<u8>),
+    AesCtr(Vec<u8>, u32),
+    RsaPss(u32),
+    RsassaPkcs1v15,
+    Ecdsa(Sha),
+    RsaOaep(Option<Vec<u8>>),
+}
+
+#[derive(Debug)]
+pub enum CryptoNamedCurve {
+    P256,
+    P384,
+}
+
+fn get_named_curve(ctx: &Ctx<'_>, curve: &str) -> Result<CryptoNamedCurve> {
+    match curve.to_ascii_uppercase().as_str() {
+        "P-256" => Ok(CryptoNamedCurve::P256),
+        "P-384" => Ok(CryptoNamedCurve::P384),
+        _ => Err(Exception::throw_message(ctx, "named_curve not found")),
+    }
+}
+
+#[derive(Debug)]
+pub enum DeriveAlgorithm {
+    Edch {
+        curve: CryptoNamedCurve,
+        public: Vec<u8>,
+    },
+    Hkdf {
+        hash: Sha,
+        salt: Vec<u8>,
+        info: Vec<u8>,
+    },
+    Pbkdf2 {
+        hash: Sha,
+        salt: Vec<u8>,
+        iterations: u32,
+    },
+}
+
+#[derive(Debug)]
+pub enum KeyGenAlgorithm {
+    Rsa {
+        modulus_length: u32,
+        public_exponent: Vec<u8>,
+    },
+    Ec {
+        curve: CryptoNamedCurve,
+    },
+    Aes {
+        length: u32,
+    },
+    Hmac {
+        hash: Sha,
+        length: Option<u32>,
+    },
+}
+
+pub async fn subtle_decrypt<'js>(
+    ctx: Ctx<'js>,
+    algorithm: Value<'js>,
+    key: ObjectBytes<'js>,
+    data: ObjectBytes<'js>,
+) -> Result<ArrayBuffer<'js>> {
+    let algorithm = extract_algorithm_object(&ctx, &algorithm)?;
+
+    let bytes = decrypt(&ctx, &algorithm, key.as_bytes(), data.as_bytes())?;
+    ArrayBuffer::new(ctx, bytes.as_slice())
```

Comment:
Do not pass slice as this will be an extra copy. Pass the vec directly to `ArrayBuffer`